### PR TITLE
BE-52_REST_methods_to_internship_controller

### PR DESCRIPTION
**Pull request by shpaser2**
[BE-52] get, post, put, delete for internships (two controllers),
    Country InternshipType Skill Subject moved to separate packages (they are not used in InternshipDetailsDto),
    posted internships - visible for interns and only get and put requests,
    unposted internships - visible for admins and get put post delete requests.

Unposted internships are visible for admins and superadmins, this internships can be updated, deleted, 
added and send to frontend. 
Posted internships can be changed or recieved on frontend.
Also it is possible to change status from published (posted) to unpablished and vise versa by seaparate get requests.

UPD 29-04-21:
1) 
for interns:
GET requests (All and by Id) on /internship url with only Published.VISIBLE_FOR_INTERNS internships.
2)
for admins:
GET requests (All and by Id) on /internship-administration url for both Published.VISIBLE_FOR_INTERNS and Published.VISIBLE_FOR_ADMINS internships.
POST requests on urls /internship-administration/registration and /internship-administration/upload
PUT requests  on url /internship-administration/{id} for only Published.VISIBLE_FOR_ADMINS internships.
DELETE requests on url /internship-administration/{id} for only Published.VISIBLE_FOR_ADMINS internships.
PATCH requests on url /internship-administration/{id} for both Published.VISIBLE_FOR_INTERNS and Published.VISIBLE_FOR_ADMINS internships.
3)
Errors messages for errors with rsql in Iternships have suggestions for better request.
4)
Header after successful POST has location (url) to the new created internship.

### CODE REVIEW

**Comment by nazfurdychka** [comment]
i have problems with internship service, for example there isn't any method entutyToDto in InternshipDetailsMapper class

**Comment by shpaser2** [comment]
> i have problems with internship service, for example there isn't any method entityToDto in InternshipDetailsMapper class, also imports are missing in mappers classes

You should try to check imports of classes Country, Subject, Skill, InternshipType and clean project with maven. It should help you.

File: src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import com.exadel.project.internship.service.InternshipService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("internship")
```
Comment (serkushner) [inline src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java:16]: /internship ?

File: src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    @GetMapping(value = ID)
+    public ResponseEntity<InternshipDetailsDTO> findPostedInternshipById(@PathVariable ("id") Long id) throws EntityNotFoundException {
+        return ResponseEntity.ok(internshipService.getPostedById(id));
+    }
+
+    @GetMapping(value = ID + "/disapprove")
+    public ResponseEntity<InternshipDetailsDTO> findAndPostUnpostedInternshipById(@PathVariable ("id") Long id) throws EntityNotFoundException {
+        return ResponseEntity.ok(internshipService.disapproveUnpostedById(id));
+    }
```
Comment (serkushner) [inline src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java:37]: Why we need such endpoint? Maybe it's better to have one generic get by specification endpoint?

File: src/main/java/com/exadel/project/internship/controller/UnpostedInternshipController.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.util.List;
+
+@RestController
+@RequestMapping("unposted-internship")
```
Comment (serkushner) [inline src/main/java/com/exadel/project/internship/controller/UnpostedInternshipController.java:19]: /unposted-internship ?

File: src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java (hunk 1 above)
Comment (shpaser2) [reply src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java]: fixed by single controller

File: src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java (hunk 2 above)
Comment (shpaser2) [reply src/main/java/com/exadel/project/internship/controller/PostedInternshipController.java]: fixed by one PATCH endpoint with one param, which should be equal to "VISIBLE_FOR_ADMINS" or "VISIBLE_FOR_INTERNS"

File: src/main/java/com/exadel/project/internship/controller/UnpostedInternshipController.java (hunk 3 above)
Comment (shpaser2) [reply src/main/java/com/exadel/project/internship/controller/UnpostedInternshipController.java]: fixed by single controller with separated url for interns and admins

**Review by serkushner — APPROVED** [review]

**Review by AlKaliada — APPROVED** [review]

**Pull request merged**